File: problems/3407-substring-matching-pattern/solution.py
class Solution:
    def hasMatch(self, s: str, p: str) -> bool:
        star_index = p.find('*')
        prefix = p[:star_index]
        suffix = p[star_index + 1:]
        for i in range(len(s) - len(p) + 1):
            if s[i:i + len(prefix)] == prefix and s[i + len(prefix):i + len(s) - len(suffix)] == s[i + len(prefix):i + len(s) - len(suffix)] and s[len(s) - len(suffix):] == s[len(s) - len(suffix):]:
                if s[len(s) - len(suffix):] == suffix:
                    return True
                if len(suffix) == 0:
                  return True
        return False

File: problems/3407-substring-matching-pattern/solution.cpp
class Solution {
public:
    bool hasMatch(string s, string p) {
        int n = s.length();
        int m = p.length();
        int starIndex = -1;
        for (int i = 0; i < m; ++i) {
            if (p[i] == '*') {
                starIndex = i;
                break;
            }
        }
        if (starIndex == -1) {
            return s.find(p) != string::npos;
        }
        for (int i = 0; i <= n - (m - 1); ++i) {
            bool match = true;
            for (int j = 0; j < starIndex; ++j) {
                if (s[i + j] != p[j]) {
                    match = false;
                    break;
                }
            }
            if (!match) continue;
            for (int j = 0; j < m - starIndex - 1; ++j) {
                if (s[i + starIndex + j + (n - (m - 1) - i)] != p[starIndex + j + 1]) {
                    match = false;
                    break;
                }
            }

            if(match)
                return true;
        }

        return false;
    }
};